wasm/wat/lexer.go

```diff
@@ -0,0 +1,239 @@
+package wat
+
+import (
+	"fmt"
+	"unicode/utf8"
+)
+
+// parseToken allows a parser to inspect a token without necessarily allocating strings
+// * source is the underlying byte stream: do not modify this
+// * tokenType is the token type
+// * beginPos is the byte position in the source where the token begins, inclusive
+// * endPos is the byte position in the source where the token ends, exclusive
+//
+// Returning an error will short-circuit any future invocations.
+type parseToken func(source []byte, tok tokenType, beginLine, beginCol, beginPos, endPos int) error
+
+// lex invokes the parser function for each token, the source is exhausted.
+//
+// Errors from the parser or during tokenization exit early, such as dangling block comments or unexpected characters.
+func lex(source []byte, parser parseToken) error {
+	length := len(source)
+	// p is the position in the source, and a parameter to the parser.
+	p := 0
+	// line is the line number in the source as determined by unescaped '\n' characters, and a parameter to the parser.
+	line := 1
+	// col is the UTF-8 aware column number, and a parameter to the parser
+	col := 0
+
+	// Block comments, ex. (; comment ;), can span multiple lines and also nest, ex. (; one (; two ;) ).
+	// There may be no block comments, but we declare the variable that tracks them here, as it is more efficient vs
+	// inline processing.
+	blockCommentLevel := 0
+
+	for ; p < length; p = p + 1 {
+		b1 := source[p]
+
+		// The spec does not consider newlines apart from '\n'. Notably, a bare '\r' is not a newline here.
+		// See https://www.w3.org/TR/wasm-core-1/#text-comment
+		if b1 == '\n' {
+			line = line + 1
+			col = 0
+			continue // next line
+		}
+
+		// Otherwise, advance the column regardless of whether we are whitespace or not
+		col = col + 1                              // the current character is at least one byte long
+		if b1 == ' ' || b1 == '\t' || b1 == '\r' { // fast path ASCII whitespace
+			continue // next whitespace
+		}
+
+		// Handle comments, noting they can be nested
+		switch b1 {
+		case '(':
+			peekPos := p + 1
+			if peekPos < length && source[peekPos] == ';' { // block comment
+				p = peekPos // continue after "(;"
+				col = col + 1
+				blockCommentLevel = blockCommentLevel + 1
+				continue
+			}
+		case ';': // possible line comment or block comment end
+			peekPos := p + 1
+			if peekPos < length {
```

Comment:
same what if peekPos < length doesn't satisfy? better reporting parse error? :D

Reply:
it will error because ';' is an invalid token. I think what's confusing you perhaps is that we are only handling whitespace indicators in this section. Anything that passes down will become a tokenization pass or fail (and ';' would fail).

It is a bit easier to keep the responsibility separate (whitespace including comment handling) from tokenization.

Reply:
sounds like I need a comment

Reply:
fixed